shared/dataflow/codeql/dataflow/test/InlineFlowTest.qll

```diff
@@ -0,0 +1,117 @@
+/**
+ * Provides a simple base test for flow-related tests using inline expectations.
+ *
+ * To add this framework to a new language, add a new file (usually called `InlineFlowTest.qll`)
+ * with:
+ * - `private import codeql.dataflow.test.InlineFlowTest`
+ * - private imports of the libraries implementing `DataFlow::InputSig`, `TaintTracking::InputSig`,
+ *   and `InlineExpectationsTest::InlineExpectationsTestSig`.
+ * - An implementation of the signature `InputSig` defined below.
+ * - An import of an appropriately instantiated `InlineFlowTestMake` module.
+ *
+ * To declare expectations, you can use the `$ hasTaintFlow` or `$ hasValueFlow` comments within the
+ * test source files. For example, in the case of Ruby `test.rb` file:
+ * ```rb
+ * s = source(1)
+ * sink(s); // $ hasValueFlow=1
+ * t = "foo" + taint(2);
+ * sink(t); // $ hasTaintFlow=2
+ * ```
+ *
+ * If you are only interested in value flow, then instead of importing `DefaultFlowTest`, you can import
+ * `ValueFlowTest<DefaultFlowConfig>`. Similarly, if you are only interested in taint flow, then instead of
+ * importing `DefaultFlowTest`, you can import `TaintFlowTest<DefaultFlowConfig>`. In both cases
+ * `DefaultFlowConfig` can be replaced by another implementation of `DataFlow::ConfigSig`.
+ *
+ * If you need more fine-grained tuning, consider implementing a test using `InlineExpectationsTest`.
+ */
+
+private import codeql.dataflow.DataFlow as DF
+private import codeql.dataflow.TaintTracking as TT
+private import codeql.util.test.InlineExpectationsTest as IET
+
+signature module InputSig<DF::InputSig DataFlowLang> {
+  predicate defaultSource(DataFlowLang::Node source);
+
+  predicate defaultSink(DataFlowLang::Node source);
+
+  string getArgString(DataFlowLang::Node src, DataFlowLang::Node sink);
+}
+
+module InlineFlowTestMake<
+  DF::InputSig DataFlowLang, TT::InputSig<DataFlowLang> TaintTrackingLang,
+  IET::InlineExpectationsTestSig Test, InputSig<DataFlowLang> Impl>
+{
+  private module DataFlow = DF::DataFlowMake<DataFlowLang>;
+
+  private module TaintTracking = TT::TaintFlowMake<DataFlowLang, TaintTrackingLang>;
+
+  private module InlineExpectationsTest = IET::Make<Test>;
+
+  module DefaultFlowConfig implements DataFlow::ConfigSig {
+    predicate isSource(DataFlowLang::Node source) { Impl::defaultSource(source) }
+
+    predicate isSink(DataFlowLang::Node sink) { Impl::defaultSink(sink) }
+
+    int fieldFlowBranchLimit() { result = 1000 }
+  }
+
+  private module NoFlowConfig implements DataFlow::ConfigSig {
+    predicate isSource(DataFlowLang::Node source) { none() }
+
+    predicate isSink(DataFlowLang::Node sink) { none() }
+  }
+
+  module FlowTest<DataFlow::ConfigSig ValueFlowConfig, DataFlow::ConfigSig TaintFlowConfig> {
```

Comment:
Can this be private?

Reply:
No, there are tests that override both configurations. For example:

https://github.com/github/codeql/blob/3343b780154c5fafe9e378412697fb4c43e5ff3e/ruby/ql/test/library-tests/frameworks/sinatra/Flow.ql#L19